src/node_buffer.cc

```diff
@@ -607,13 +607,37 @@ int32_t IndexOf(const char* haystack,
                 const char* needle,
                 size_t n_length) {
   CHECK_GE(h_length, n_length);
-  // TODO(trevnorris): Implement Boyer-Moore string search algorithm.
-  for (size_t i = 0; i < h_length - n_length + 1; i++) {
-    if (haystack[i] == needle[0]) {
-      if (memcmp(haystack + i, needle, n_length) == 0)
-        return i;
-    }
+
+  const char* inc;
+  const void* ptr;
+
+  // Guard against zero length searches
+  if (h_length == 0 || n_length == 0 || n_length > h_length)
+    return -1;
+
+  // Single character case
+  if (n_length == 1) {
+    ptr = memchr(haystack, *needle, h_length);
+    if (ptr == nullptr)
+      return -1;
+    return static_cast<int32_t>(static_cast<const char*>(ptr) - haystack);
   }
+
+  // Do multicharacter string match
+  inc = haystack;
+  do {
+    ptr = memchr(inc, *needle, h_length - static_cast<int32_t>(inc - haystack));
+    if (ptr != nullptr) {
+      const void* ptr2 = memmem(inc,
```

Comment:
Maybe just pick a "good" algorithm from [here](http://www-igm.univ-mlv.fr/~lecroq/string/index.html) to use on Windows (or everywhere for consistency?). Start with [Turbo Boyer-Moore](http://www-igm.univ-mlv.fr/~lecroq/string/node15.html) perhaps?